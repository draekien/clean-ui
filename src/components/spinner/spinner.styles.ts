import { Theme, ThemeUIStyleObject } from 'theme-ui';
import { alpha } from '@theme-ui/color';

export const overlayCss: ThemeUIStyleObject = {
  position: 'fixed',
  display: 'flex',
  top: 0,
  right: 0,
  bottom: 0,
  left: 0,
  backgroundColor: (t: Theme) => `${alpha('muted', 0.5)(t)}`,
  zIndex: 'overlay',
  alignItems: 'center',
  justifyContent: 'center',
};
