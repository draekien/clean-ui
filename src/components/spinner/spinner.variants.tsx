// eslint-disable-next-line tsdoc/syntax
/** @jsxImportSource theme-ui */
import React from 'react';
import { Size } from '../../types/layouts';
import * as styles from './spinner.variants.styles';

export interface SpinnerVariantProps {
  size: Size;
}

export const mapSizeToRem = (size: Size) => {
  if (size === 'small') return '1rem';
  if (size === 'medium') return '2rem';
  return '3rem';
};

const spinnerId = 'clean_ui_spinner';

const Donut: React.FC<SpinnerVariantProps> = ({ size }) => {
  return <div id={spinnerId} sx={styles.donutSpinnerCss({ size })} />;
};

const Spiral: React.FC<SpinnerVariantProps> = ({ size }) => {
  return <div id={spinnerId} sx={styles.spiralSpinnerCss({ size })} />;
};

const Bars: React.FC<SpinnerVariantProps> = ({ size }) => {
  return (
    <div id={spinnerId} sx={styles.barsAndDotsContainerCss}>
      {[1, 2, 3].map((val) => (
        <div key={val} sx={styles.barsSpinnerCss({ size })} />
      ))}
    </div>
  );
};

const Dots: React.FC<SpinnerVariantProps> = ({ size }) => {
  return (
    <div id={spinnerId} sx={styles.barsAndDotsContainerCss}>
      {[1, 2, 3].map((val) => (
        <div key={val} sx={styles.dotsSpinnerCss({ size })} />
      ))}
    </div>
  );
};

export const SpinnerVariants = { Donut, Spiral, Bars, Dots };
