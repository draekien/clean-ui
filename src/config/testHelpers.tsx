import React, { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: ReactNode) =>
  render(<ThemeProvider>{element}</ThemeProvider>);
