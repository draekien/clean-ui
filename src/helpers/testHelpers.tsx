import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: React.ReactNode) => {
  return render(<ThemeProvider>{element}</ThemeProvider>);
};

export const hexToRgb = (hex: string) => {
  hex = hex.replace(/[^0-9A-F]/gi, '');
  const bigint = parseInt(hex, 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
};
