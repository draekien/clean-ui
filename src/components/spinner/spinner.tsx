/** @jsxImportSource theme-ui */
import React from 'react';
import { Size } from '../../types/layouts';
import { SpinnerVariants } from './spinner.variants';
import { Overlay } from '../overlay/overlay';

export type SpinnerVariant = 'donut' | 'dots' | 'bars' | 'spiral';

export interface SpinnerProps {
  /** the type of spinner to render
   * @default 'donut'
   */
  variant?: SpinnerVariant;
  /** the size of the spinner to render
   * @default 'medium'
   */
  size?: Size;
  /** should the spinner take up the whole page?
   * @default false
   */
  fullPage?: boolean;
}

export const Spinner: React.FC<SpinnerProps> = ({
  variant = 'donut',
  size = 'medium',
  fullPage = false,
}) => {
  const Spinner = () => {
    if (variant === 'spiral') {
      return <SpinnerVariants.Spiral size={size} />;
    }

    if (variant === 'bars') {
      return <SpinnerVariants.Bars size={size} />;
    }

    if (variant === 'dots') {
      return <SpinnerVariants.Dots size={size} />;
    }

    // default to donut
    return <SpinnerVariants.Donut size={size} />;
  };

  return fullPage ? (
    <Overlay>
      <Spinner />
    </Overlay>
  ) : (
    <Spinner />
  );
};
