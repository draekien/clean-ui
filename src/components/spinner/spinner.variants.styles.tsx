import { ThemeUIStyleObject } from 'theme-ui';
import { keyframes } from '@emotion/react';
import { mapSizeToRem, SpinnerVariantProps } from './spinner.variants';

const rotation = keyframes`0% {
  transform: rotate(0deg);
}
100% {
  transform: rotate(360deg);
}`;

export const donutSpinnerCss = ({ size }: SpinnerVariantProps): ThemeUIStyleObject => {
  const css: ThemeUIStyleObject = {
    width: mapSizeToRem(size),
    height: mapSizeToRem(size),
    border: '0.25rem solid',
    borderColor: 'muted',
    borderBottomColor: 'primary',
    borderRadius: 'max',
    display: 'inline-flex',
    animation: `${rotation} 1s linear infinite`,
  };

  return css;
};

const spiral = keyframes`50% { transform: scale(1) translate(-50%, -50%)}`;

export const spiralSpinnerCss = ({ size }: SpinnerVariantProps): ThemeUIStyleObject => {
  const sizeInRem = mapSizeToRem(size);

  const css: ThemeUIStyleObject = {
    width: sizeInRem,
    height: sizeInRem,
    display: 'inline-flex',
    margin: '1rem auto',
    position: 'relative',
    color: 'muted',
    animation: `${rotation} 1s linear infinite`,
    '::before, ::after': {
      content: '""',
      position: 'absolute',
      width: '50%',
      height: '50%',
      top: '50%',
      left: '50%',
      transform: 'scale(0.5) translate(0,0)',
      backgroundColor: 'secondary',
      borderRadius: 'max',
      animation: `${spiral} 1s infinite ease-in-out`,
    },
    '::before': {
      backgroundColor: 'primary',
      transform: `scale(0.5) translate(-${sizeInRem}, -${sizeInRem})`,
    },
  };

  return css;
};

export const barsAndDotsContainerCss: ThemeUIStyleObject = {
  display: 'inline-flex',
  '> div:nth-of-type(1)': {
    animationDelay: '0',
    marginLeft: '0',
  },
  '> div:nth-of-type(2)': {
    animationDelay: '0.09s',
  },
  '> div:nth-of-type(3)': {
    animationDelay: '0.18s',
  },
};

const loadingBars = keyframes`
0% {
  transform: scale(1);
}

20% {
  transform: scale(1, 2);
}

40% {
  transform: scale(1);
}`;

export const barsSpinnerCss = ({ size }: SpinnerVariantProps): ThemeUIStyleObject => {
  const css: any = {
    animation: `${loadingBars} 1s ease-in-out infinite`,
    backgroundColor: 'primary',
    borderRadius: 'sm',
    display: 'inline-flex',
  };

  if (size === 'small') {
    css.height = '0.75rem';
    css.width = '0.125rem';
    css.ml = '0.125rem';
  } else if (size === 'medium') {
    css.height = '1.25rem';
    css.width = '0.25rem';
    css.ml = '0.25rem';
  } else {
    css.height = '1.75rem';
    css.width = '0.375rem';
    css.ml = '0.375rem';
  }

  return css;
};

const loadingDots = keyframes`
0% {
  transform: translateY(0);
}

20% {
  transform: translateY(-75%);
}

40% {
  transform: translateY(0);
}`;

export const dotsSpinnerCss = ({ size }: SpinnerVariantProps): ThemeUIStyleObject => {
  const css: any = {
    animation: `${loadingDots} 1s ease-in-out infinite`,
    backgroundColor: 'primary',
    display: 'inline-flex',
    borderRadius: 'max',
    ':nth-of-type(1)': { backgroundColor: 'p-400' },
    ':nth-of-type(2)': { backgroundColor: 'p-300' },
    ':nth-of-type(3)': { backgroundColor: 'p-200' },
  };

  if (size === 'small') {
    css.height = '0.25rem';
    css.width = '0.25rem';
    css.ml = '0.125rem';
  } else if (size === 'medium') {
    css.height = '0.5rem';
    css.width = '0.5rem';
    css.ml = '0.25rem';
  } else {
    css.height = '0.75rem';
    css.width = '0.75rem';
    css.ml = '0.375rem';
  }

  return css;
};
