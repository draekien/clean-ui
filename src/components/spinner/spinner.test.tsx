import React from 'react';
import { themedRender } from '../../helpers/testHelpers';
import { SpinnerVariant, Spinner } from './spinner';

describe('Spinner Component', () => {
  const variants: SpinnerVariant[] = ['bars', 'donut', 'dots', 'spiral'];

  variants.map((variant) => {
    it(`${variant} variant should match snapshot`, () => {
      const { container } = themedRender(<Spinner variant={variant} />);

      expect(container).toMatchSnapshot();
    });
  });
});
