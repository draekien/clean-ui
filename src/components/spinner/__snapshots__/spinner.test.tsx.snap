// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner Component bars variant should match snapshot 1`] = `
<div>
  <div
    class="css-5bicxv-Bars"
  >
    <div
      class="css-1a9qa0f"
    />
    <div
      class="css-1a9qa0f"
    />
    <div
      class="css-1a9qa0f"
    />
  </div>
</div>
`;

exports[`Spinner Component donut variant should match snapshot 1`] = `
<div>
  <div
    class="css-1luozaz-Donut"
  />
</div>
`;

exports[`Spinner Component dots variant should match snapshot 1`] = `
<div>
  <div
    class="css-10j4viu-Dots"
  >
    <div
      class="css-1crq1ew"
    />
    <div
      class="css-1crq1ew"
    />
    <div
      class="css-1crq1ew"
    />
  </div>
</div>
`;

exports[`Spinner Component spiral variant should match snapshot 1`] = `
<div>
  <div
    class="css-enyyst-Spiral"
  />
</div>
`;
